src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,357 @@
+
+            // The indices in indexToEdit kinds are not guaranteed to be in chronological order when we
+            // extract them from the dictionary. We must sort the edit kinds by index since we need to
+            // know what kind of edits surround a given index in order to potentially combine them into
+            // one edit.
+            Array.Sort(editIndices);
+
+            // Example to give clarity to orderedEditNumber and orderedTokenNumber variables defined below:
+            // Non-grouped semantic tokens: 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14
+            // orderedEditNumber:           0                   1
+            // orderedTokenNumber:          0         1         2
+            for (var orderedEditNumber = 0; orderedEditNumber < editIndices.Length; orderedEditNumber++)
+            {
+                var orderedTokenNumber = editIndices[orderedEditNumber];
+                var initialEditKind = indexToEditKinds[orderedTokenNumber];
+
+                if (initialEditKind == SemanticTokenEditKind.Update)
+                {
+                    orderedEditNumber = AddUpdateEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndices, orderedEditNumber,
+                        groupedSemanticToken: newGroupedSemanticTokens[orderedTokenNumber],
+                        editStartPosition: orderedTokenNumber * 5);
+                }
+                else if (initialEditKind == SemanticTokenEditKind.Insert)
+                {
+                    orderedEditNumber = AddInsertionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndices, orderedEditNumber,
+                        groupedSemanticToken: newGroupedSemanticTokens[orderedTokenNumber],
+                        editStartPosition: orderedTokenNumber * 5);
+                }
+                else
+                {
+                    Contract.ThrowIfFalse(initialEditKind == SemanticTokenEditKind.Delete);
+                    orderedEditNumber = AddDeletionEdit(
+                        indexToEditKinds, semanticTokensEdits, editIndices, orderedEditNumber,
+                        editStartPosition: orderedTokenNumber * 5);
+                }
+            }
+
+            return semanticTokensEdits.ToArray();
+
+            // Local functions
+            static int AddUpdateEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int i,
+                SemanticToken groupedSemanticToken,
+                int editStartPosition)
+            {
+                var deleteCount = 5;
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                tokensToInsert.AddRange(groupedSemanticToken.ConvertToArray());
```

Comment:
still allocating :)  do:

groupedSemanticToken.AddToEnd(tokensToInsert)

and have AddToEnd do the adding in a non-allocating fashion.